src/typecheck/mod.rs

```diff
@@ -1000,23 +1003,192 @@ fn walk_rrows<L: Linearizer>(
+                        row,
+                        tail: Box::new(tail),
+                    })
+                });
+
+                UnifType::Concrete(TypeF::Record(rrows))
+            }
+            Destruct::Empty => state.table.fresh_type_uvar(),
+        }
+    }
+
+    /// Extend `env` with any new bindings brought into scope in `pat`. The
+    /// types of these bindings will be inferred from `pat_ty`.
+    ///
+    /// For example, if `pat` represents the pattern `{ a, ..rest }` and
+    /// `pat_ty` is `{ a : Num, b : Str }` then the `env` will be extended
+    /// with `a : Num` and `rest : { b : Str }`.
+    pub fn inject_pattern_variables(
+        state: &State,
+        env: &mut Environment,
+        pat: &Destruct,
+        pat_ty: UnifType,
+    ) {
+        let pat_ty = pat_ty.into_root(state.table);
+
+        match pat {
+            Destruct::Record { matches, rest, .. } => {
+                let mut type_map = RecordTypes::from(&pat_ty);
+
+                matches.iter().for_each(|m| match m {
+                    Match::Simple(id, ..) => {
+                        let ty = type_map.get_type(id);
+                        env.insert(*id, ty);
+                    }
+                    Match::Assign(id, _, (bind_id, pat)) => {
+                        let ty = type_map.get_type(id);
+                        let id = bind_id.as_ref().unwrap_or(id);
+                        env.insert(*id, ty.clone());
+                        // A non-empty `pat` here means we have a nested destructuring,
+                        // so we recursively call this function.
+                        if !pat.is_empty() {
+                            inject_pattern_variables(state, env, pat, ty)
+                        }
+                    }
+                });
+
+                if let Some(id) = rest {
+                    let rest_ty = type_map.rest();
+                    env.insert(*id, rest_ty);
+                }
+            }
+            Destruct::Empty => (),
         }
-        matches.iter().for_each(|m| match m {
-            Match::Simple(id, ..) => env.insert(*id, UnifType::Concrete(TypeF::Dyn)),
-            Match::Assign(id, _, (bind_id, pat)) => {
-                let id = bind_id.as_ref().unwrap_or(id);
-                env.insert(*id, UnifType::Concrete(TypeF::Dyn));
-                if !pat.is_empty() {
-                    inject_pat_vars(pat, env);
+    }
+
+    /// A map of identifiers in a destructured record to their types.
+    ///
+    /// This allows us to be resilient to ordering differences between the
+    /// pattern and its type. As well as keeping track of which identifiers
+    /// have already been "used" in the pattern, to ensure that we can
+    /// correctly construct the type of a `..rest` match, if it exists.
+    enum RecordTypes {
+        Rows {
+            known_types: HashMap<Ident, UnifType>,
+            tail: UnifRecordRows,
+        },
+        Dyn,
```

Comment:
I'm not sure to see why we need a `Dyn`, and in which case does it arise. Shouldn't we always infer a record type from a pattern? If yes, this state shouldn't be representable, and `RecordType` should probably implement `try_from`, or `From<RecordRows>`.

Reply:
We need `Dyn` here in order to introduce the pattern variables when `walk`ing. In that case, we pass `GenericUnifType::Concrete(Dyn)` as the pattern type, and so we create a `RecordType::Dyn` which then always returns `Dyn` from `get_type` and `{ _ : Dyn }` from `rest`.

I'm totally open to better ways of structuring this if you have suggestions.

Reply:
I think you want the elaborate the pattern type even when walking. For example, during walking:

```
nickel> let r = {x = 1} in
let x | Num = r.x in
(x + 2 : Num)
3

nickel>
```

`x` is still affected the type `Num`, because it has a contract annotation. However, if we use destructuring:

```
nickel> let r = {x = 1} in
let {x | Num} = r in
(x + 2 : Num)
error: incompatible types
  ┌─ repl-input-1:3:2
  │
3 │ (x + 2 : Num)
  │  ^ this expression
  │
  = The type of the expression was expected to be `Num`
  = The type of the expression was inferred to be `Dyn`
  = These types are not compatible
```

However, the second example is desugared to the same thing as the first one, and contains as much information. The type is not really inference per se, but mostly building a map from identifiers to types based on annotations alone, which you do even during the walk phase.

Reply:
Ah yeah, this is a great point. Thank you!